### PR TITLE
Add upstream proxy authentication

### PR DESCRIPTION
**Pull request by xhy940801**
Finished #644

### CODE REVIEW

**Comment by mhils** [comment]
This looks good, thanks. Can you please add a test that verifies that the feature working? :smiley:

**Comment by xhy940801** [comment]
I've added some test cases, but I'm not sure it's enough:disappointed_relieved:

File: test/mitmproxy/test_cmdline.py [hunk 1]
```diff
@@ -53,6 +54,12 @@ def test_parse_server_spec():
         "http://")
 
 
+def test_parse_upstream_auth():
+    tutils.raises("Invalid upstream auth specification", cmdline.parse_upstream_auth, "")
+    assert cmdline.parse_upstream_auth(
+        "test:test") == "Basic" + " " + base64.b64encode("test:test")
```
Comment (Kriechi) [inline test/mitmproxy/test_cmdline.py]: I'm not sure what the RFC says, but are these allowed?
  * `test:` (empty password, maybe valid)
  * `:test` (empty username, probably invalid)
  * `:` (both empty - is it then a username by default, with empty password? or the other way around)

We should make sure to cover these cases: in the tests as well as the cli-parsing part.

**Comment by Kriechi** [comment]
Thanks! :tada:

**Comment by JakeAustwick** [comment]
I'm trying to use this right now, but hitting some issues.

```
mitmproxy -U http://xxxx.com:23333 --upstream-auth=xxx:yyy
```

Then when I try and use the proxy:
```
>>> proxies = {'http': 'http://127.0.0.1:8080'}
>>> url = 'http://geoip.nekudo.com/api'
>>> r = requests.get(url, proxies=proxies)
```

The mitmproxy console shows:
```
mitmproxy has crashed! call last):
Please lodge a bug report at: https://github.com/mitmproxy/mitmproxy, line 121, in handle
    root_layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/proxy/modes/http_proxy.py", line 25, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/tls.py", line 358, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http1.py", line 67, in __call__
    layer()
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/http.py", line 187, in __call__
    self.data.headers["Proxy-Authorization"] = self.config.upstream_auth
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/base.py", line 81, in __getattr__
    return getattr(self.ctx, name)
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/base.py", line 81, in __getattr__
    return getattr(self.ctx, name)
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/base.py", line 81, in __getattr__
    return getattr(self.ctx, name)
  File "/Library/Python/2.7/site-packages/mitmproxy/protocol/base.py", line 81, in __getattr__
    return getattr(self.ctx, name)
AttributeError: 'RootContext' object has no attribute 'data'
```

It's not the proxy because using the proxy directly with `requests` works. Any ideas?

**Comment by mhils** [comment]
@JakeAustwick: fixed, thanks.

**Comment by JakeAustwick** [comment]
Awesome. What's the easiest way for me to install this patch locally, so that I can test? I'm guessing it won't be pushed to PyPI for a while?

**Pull request merged**